### PR TITLE
Shipping address

### PR DESCRIPTION
**Pull request by qkul**

### CODE REVIEW

File: ecommers/lib/core/common/validator/shipping_address_validator.dart [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+      fullName != null && fullName.length > 3;
+
+  static bool isValidAddress(String address) =>
+      address != null && address.length > 2;
+
+  static bool isValidCity(String city) =>
+      city != null && city.length > 2;
+
+  static bool isValidState(String state) =>
+      state != null && state.length > 3;
```
Comment (Platton96) [inline ecommers/lib/core/common/validator/shipping_address_validator.dart:13]: can you check to work state?.lenght>3;

File: ecommers/lib/core/app_services/shipping_address_service.dart [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    try {
+      final uuid = Uuid();
+      shippingAddressModel.id = uuid.v1();
+      await shippingAddressRepository.add(
+        shippingAddressModel,
+      );
+
+      return shippingAddressModel;
+    } catch (ex) {
+      return null;
```
Comment (fryette) [inline ecommers/lib/core/app_services/shipping_address_service.dart]: please use logger here

File: ecommers/lib/core/provider_models/shipping_address_provider_model.dart [hunk 3]
```diff
@@ -0,0 +1,100 @@
+import 'package:ecommers/core/common/validator/index.dart';
+import 'package:ecommers/core/models/index.dart';
+import 'package:ecommers/shared/dependency_service.dart';
+import 'package:flutter/material.dart';
+
+class ShippingAddressProviderModel extends ChangeNotifier {
+  bool isValidFullName = true,
```
Comment (fryette) [inline ecommers/lib/core/provider_models/shipping_address_provider_model.dart]: why default value is true?

File: ecommers/lib/core/provider_models/shipping_address_provider_model.dart [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      isValidAddress = true,
```
Comment (fryette) [inline ecommers/lib/core/provider_models/shipping_address_provider_model.dart]: a better name is isAddressValid, please rename all props

File: ecommers/lib/core/provider_models/shipping_address_provider_model.dart [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    if (_selectedShippingAddress == null) {
+      _selectedShippingAddress = shippingAddressModel..isSelected = true;
+      shippingAddressService.editShippingAddresses(_selectedShippingAddress);
+      notifyListeners();
+    }
+
+    if (shippingAddressModel.id == _selectedShippingAddress.id) return;
+
+    shippingAddressModel.isSelected = true;
+    _selectedShippingAddress.isSelected = false;
```
Comment (fryette) [inline ecommers/lib/core/provider_models/shipping_address_provider_model.dart]: ????
what is it?

Comment (fryette) [reply ecommers/lib/core/provider_models/shipping_address_provider_model.dart]: need to discuss

File: ecommers/lib/ui/pages/add_shipping_address_page.dart [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+  }
+
+  @override
+  void dispose() {
+    disposeControllers();
+    provider.clean();
+    super.dispose();
+  }
+
+  void disposeControllers() {
```
Comment (fryette) [inline ecommers/lib/ui/pages/add_shipping_address_page.dart]: private?

File: ecommers/lib/ui/pages/add_shipping_address_page.dart [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    _zipCodeController.dispose();
+    _countryController.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (widget.shippingAddres != null) {
+      provider?.shippingAddress = widget.shippingAddres;
+    }
+    final localization = I18n.of(context);
```
Comment (fryette) [inline ecommers/lib/ui/pages/add_shipping_address_page.dart]: please remove it and use without i18n reference, just use localization from dependencies

File: ecommers/lib/ui/pages/add_shipping_address_page.dart [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    return Scaffold(
+      backgroundColor: BrandingColors.pageBackground,
+      appBar: AppBar(
+        leading: const BackButton(),
+        title: Center(
+          child: Text(
+            localization.addingShippingAddress,
+            style: Theme.of(context)
```
Comment (fryette) [inline ecommers/lib/ui/pages/add_shipping_address_page.dart]: please remove and use textTheme from dependencies

File: ecommers/lib/ui/pages/checkout_page.dart [hunk 9]
```diff
@@ -190,30 +189,50 @@ class _CheckoutPageState extends State<CheckoutPage> {
...
+      children: [
+        Text(
+          provider?.selectedShippingAddress?.fullName,
+          style: Theme.of(context)
+              .textTheme
+              .bodyText1
+              .copyWith(fontWeight: FontWeight.w600),
+        ),
+        Text(
+          '${provider?.selectedShippingAddress?.country}, ${provider?.selectedShippingAddress?.city}, ${provider?.selectedShippingAddress?.state}',
```
Comment (fryette) [inline ecommers/lib/ui/pages/checkout_page.dart]: provider here should not be null

File: ecommers/lib/ui/pages/shipping_address_page.dart [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+class _ShippingAddressState extends State<ShippingAddressPage> {
+  GlobalKey<AnimatedListState> _listKey;
+  ShippingAddressProviderModel _provider;
+  @override
+  Widget build(BuildContext context) {
+    _listKey = GlobalKey();
+    _provider = Provider.of<ShippingAddressProviderModel>(context);
+    final localization = I18n.of(context);
```
Comment (fryette) [inline ecommers/lib/ui/pages/shipping_address_page.dart]: remove and use just localization

**Review by fryette — CHANGES_REQUESTED** [review]

File: ecommers/lib/core/common/validator/shipping_address_validator.dart (hunk 1 above)
Comment (qkul) [reply ecommers/lib/core/common/validator/shipping_address_validator.dart:13]: the ' > ' method cannot be called on null

File: ecommers/lib/ui/pages/shipping_address_page.dart (hunk 10 above)
Comment (qkul) [reply ecommers/lib/ui/pages/shipping_address_page.dart]: done

File: ecommers/lib/ui/pages/add_shipping_address_page.dart (hunk 6 above)
Comment (qkul) [reply ecommers/lib/ui/pages/add_shipping_address_page.dart]: done

File: ecommers/lib/core/provider_models/shipping_address_provider_model.dart (hunk 3 above)
Comment (qkul) [reply ecommers/lib/core/provider_models/shipping_address_provider_model.dart]: fixed

File: ecommers/lib/core/provider_models/shipping_address_provider_model.dart (hunk 5 above)
Comment (qkul) [reply ecommers/lib/core/provider_models/shipping_address_provider_model.dart]: fixed

File: ecommers/lib/core/provider_models/index.dart [hunk 11]
```diff
@@ -10,3 +10,6 @@ export 'products_grid_provder_model.dart';
 export 'search_page_provider_model.dart';
 export 'search_query_provider_model.dart';
 export 'shell_provider_model.dart';
+export 'shipping_address_provider_model.dart';
+
```
Comment (Platton96) [inline ecommers/lib/core/provider_models/index.dart]: remove useless space

File: ecommers/lib/core/provider_models/shipping_address_provider_model.dart [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+      await shippingAddressService
+          .updateSelectedShippingAddress(shippingAddress);
+      selectedShippingAddress[0] = shippingAddress;
+      notifyListeners();
+    }
+  }
+
+  bool _isValidShippingAddress() {
+    isFullNameValid =
+        ShippingValidator.isValidFullName(shippingAddress.fullName);
```
Comment (Platton96) [inline ecommers/lib/core/provider_models/shipping_address_provider_model.dart]: different approach naming isFullNameValid  isValidFullName(shippingAddress.fullName); Use one approach

**Review by Platton96 — DISMISSED** [review]

File: ecommers/lib/core/provider_models/shipping_address_provider_model.dart (hunk 12 above)
Comment (qkul) [reply ecommers/lib/core/provider_models/shipping_address_provider_model.dart]: renamed

**Review by fryette — APPROVED** [review]

File: ecommers/lib/core/provider_models/index.dart (hunk 11 above)
Comment (qkul) [reply ecommers/lib/core/provider_models/index.dart]: removed

**Pull request merged**